File: haskell/glue/src/Glue/Lib/Math/Utility/Trunc.hs
module Glue.Lib.Math.Utility.Trunc where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Truncate function
-- Mirrors Haskell Glue.Lib.Math.Utility.Trunc.trunc exactly
trunc :: IR Eval
trunc = NativeFunc truncImpl

-- Truncate function implementation
-- Mirrors Haskell Glue.Lib.Math.Utility.Trunc.truncImpl exactly
truncImpl :: IR Eval -> Eval (IR Eval)
truncImpl arg = case arg of
    Integer n -> pure $ Integer n
    Float n -> pure $ Integer (Prelude.truncate n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/utility/trunc.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Truncate function (removes decimal part)
/// Mirrors Haskell Glue.Lib.Math.Utility.Trunc.trunc exactly
final Ir trunc = IrNativeFunc(truncImpl);

/// Truncate function implementation (removes decimal part)
/// Mirrors Haskell Glue.Lib.Math.Utility.Trunc.truncImpl exactly
Eval<Ir> truncImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(IrInteger(n)),
    IrFloat(value: final n) => Eval.pure(IrInteger(n.truncate())),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
